src/stats/man/dprime.1
.TH DPRIME 1 "August 21, 1990" "\(co 1980 Gary Perlman" "|STAT" "UNIX User's Manual"
.SH NAME
dprime \- compute d' and beta for signal detection data
.SH SYNOPSIS
.B dprime
[-p] [hit\-rate\ false-alarm-rate] [hits\ false\-alarms\ misses\ correct\-rejections]
.SH DESCRIPTION
.I dprime
works from data collected in a signal detection task in which
a signal is either present or not in a stimulus,
and an observer either indicates that a signal was detected or not.
.nf

.if n .ta 20n +20n +20n +20n
.if t .ta 1.5i +1.5i +1.5i +1.5i
	Signal Presented?
	Yes	No
Signal Detected?
Yes	hit	fa
No	miss	cr

.fi
In the table above, a
.I hit
is a signal that was presented and detected, a
.I miss
is a signal that was presented but not detected, a
.I fa
(for false alarm)
is a signal that was not presented but mistakenly detected, and a
.I cr
(for correct rejection)
is a signal that was not presented and not reported as detected by the observer.
The hit rate, sometimes called
.I hr,
is the percentage of signals that were detected
and is based on the number of hits and misses:
.ce
hr = #hit / (#hit + #miss).
The false alarm rate, sometimes called
.I far,
is  the percentage of non-signals that were reported as signals
and is based on the number of false-alarms and correct-rejections:
.ce
far = #fa / (#fa + #cr).
.PP
.I dprime
reads its input in several formats.
.I dprime
can be given two arguments: the hit-rate and the false-alarm-rate,
for which
.I dprime
will compute d', beta, and the criterion.
.I dprime
can also be give four arguments: the number of hits, the number of false-alarms,
the number of misses and the number of correct-rejections,
for which
.I dprime
will compute the hit-rate and false-alarm-rate and
then compute d', beta, and the criterion.
Otherwise, if no values are provided on the command line,
.I dprime
reads raw data from the standard input.
If raw data are input,
.I dprime
assumes a two column input in which the first column tells whether
signal+noise or just noise were presented,
and the second column tells how the observer responded.
The following strings can be used to indicate affirmative answers
.ce
signal, yes, 1
while the following can be used to indicate negative:
.ce
noise, no, 0
Upper case forms for the above are allowed.
When reading the standard input,
.I dprime
allows an optional third column that gives a repetition count
for the case.
.SH OPTIONS
.de OP
.TP
.B -\\$1 \\$2
..
.OP p
print the `precision' of the observer, calculated as:
.ce
#hit / (#hit + #false-alarm)
In the information retrieval field,
it is common to evaluate the effectiveness of a retrieval with two values:
.nf
  R = recall = proportion of relevant items retrieved (the hit-rate)
  P = precision = proportion of retrieved items that are relevant
.fi
The correspondence to signal detection is that `relevance' is the
signal to be detected and has been discussed in
Swets, J. A.  (1969)
.ft I
Effectiveness of Information Retrieval Methods.
.ft B
American Documentation, 20,
.ft R
72-89.
.SH ALGORITHM
.PP
The value for d' is the Z value of the hit-rate
minus that of the false-alarm-rate.
.ce
d' = Z(hr) - Z(far)
This reflects the distance between the two distributions:
signal, and signal+noise.
Though Z values can have any real value,
normally distributed ones are between -2 and 2 about 95% of the time,
so differences of twice that would be rare.
.PP
The value for beta is the ratio of the normal density functions at
the criterion of the Z values used in the computation of d'.
This reflects an observer's bias to say `yes' or `no'
with the unbiased observer having a value around 1.0.
As the bias to say `yes' increases,
resulting in a higher hit-rate and false-alarm-rate,
beta approaches 0.0.
As the bias to say `no' increases,
resulting in a lower hit-rate and false-alarm-rate,
beta increases over 1.0 on an open-ended scale.
A major reason for doing a signal detection analysis is to get a measure
of discrimination that is constant over observer biases,
but the invariance of d' over different beta is often not certain.
The `criterion' is the distance from the mean of the noise (non-signal)
distribution to the point above which the observer says `yes'
and is in the same units as d'.
.PP
If results indicate a perfect observer (e.g., all hits and no misses),
then then d' would be infinite.
In realistic conditions, this would never happen.
In such cases, d' is reported as 0.0, and beta is reported as 1.0.
.SH EXAMPLE
.nf
.ta .75i +.75i +.75i +.75i +.75i
dprime .7 .4     # will print
hr	far	dprime	beta	criterion	
0.700	0.400	0.778	0.900	0.253	

dprime 70 400 30 600 # will print the same, but with a table
.fi
.SH REFERENCE
Coombs, C. H., Dawes, R. M., and Tversky, A. (1970)
.ft B
Mathematical Psychology,
.ft R
Englewood Cliffs, NJ: Prentice-Hall.
See Chapter 6 on
.ft I
The Theory of Signal Detectability,
.ft R
pp. 165-201.

Salton, G, & McGill, M. J. (1983)
.ft B
Introduction to Modern Information Retrieval,
.ft R
New York: McGraw-Hill.
See Chapter 5 on
.ft I
Retrieval Evaluation,
especially pp. 164-184.
.SH BUGS
The program has not been tested extensively.


src/stats/src/dprime.c
/*  Copyright 1982 Gary Perlman */

#include "stat.h"
PGM(dprime,Signal Detection Theory Analysis,5.3,08/22/90)


/*HISTORY
	08/29/90	GP	moved hit rate and false alarm rate over to line up
	08/29/90	GP	changed counts to longs for huges sample sizes
	08/22/90	GP	added checks for == 0.0, == 1.0 for _rate values
	08/22/90	GP	added criterion reporting
	08/21/90	GP	added -p option to print precision
	08/21/90	GP	added optional repetition count field for file input
	08/21/90	GP	added raw frequency count option for command line
	08/21/90	GP	found bug in yesno: always returned "yes"
*/

Boole	InfoVersion;          /* print version information */
Boole	InfoLimits;           /* print program limits */
Boole	InfoOptions;          /* print usage information */
Boole	Doprecision;          /* print IR computation of Hits/(Hits+FA) */

#define	SQRT2PI               2.506628274631000502415765
#define	normaldensity(z)      (exp (-z*z * .5) / SQRT2PI)
double	critz ();

#define	MAXCOLS	      3
#define MAXCHARS BUFSIZ           /* maximum number of chars in lines */
char	*Column[MAXCOLS];
typedef	long	Count;
#define	atoc(s) atol(s)
long 	atol();
Count	Hit, Miss, False_alarm, Correct_rejection;
double	Hit_rate, False_alarm_rate, Precision;

fparselin (ioptr, array, maxcols)
FILE	*ioptr;
char	**array;
	{
	static	char	line[MAXCHARS];
	if (fgets (line, sizeof (line), ioptr))
		return (parselin (line, array, maxcols));
	else
		return (EOF);
	}

main (argc, argv) char **argv;
	{
	int 	ncols;
	int 	optind;
	int 	nargs;
	Boole 	presented, detected;   /* was signal presented? detected? */
	Count	count;                 /* repetition count of file input line */

	ARGV0;
	optind = initial (argc, argv);
	nargs = argc - optind;

	if (nargs == 0)
		{
		checkstdin ();
		while ((ncols = fparselin (stdin, Column, MAXCOLS)) != EOF)
			{
			if (ncols == 0)
				continue;
			count = 1;
			presented = yesno (Column[0]);
			detected  = yesno (Column[1]);
			if (ncols != 2) /* should allow optional count as third field */
				{
				if (ncols == 3 && isinteger (Column[2]))
					{
					count = atoc (Column[2]);
					if (count <= 0)
						ERRMSG1 (repetition count in input (%s) must be positive, Column[2])
					}
				else
					ERRMSG0 (each line must have 2 columns with optional repetition count)
				}
			if (presented)
				{
				if (detected)
					Hit += count;
				else
					Miss += count;
				}
			else /* not presented */
				{
				if (detected)
					False_alarm += count;
				else
					Correct_rejection += count;
				}
			}
		rawtable ();
		}
	else if (nargs == 2) /* hit rate and false alarm rate supplied */
		{
		if (Doprecision)
			{
			WARNING (precision can not be computed from hit-rate and false-alarm rate)
			Doprecision = FALSE;
			}
		if (!number (argv[optind]))
			ERRNUM (argv[optind],hit-rate)
		if (!number (argv[optind+1]))
			ERRNUM (argv[optind+1],false-alarm-rate)
		Hit_rate = atof (argv[optind]);
		False_alarm_rate = atof (argv[optind+1]);
		}
	else /* #hits #fa #miss #cr on command line */
		{
		if (!isinteger (argv[optind+0]) || (Hit = atoc (argv[optind+0])) < 0)
			ERRMSG1 (hits (%s) is not a valid count, argv[optind+0])
		if (!isinteger (argv[optind+1]) || (False_alarm = atoc (argv[optind+1])) < 0)
			ERRMSG1 (false-alarms (%s) is not a valid count, argv[optind+1])
		if (!isinteger (argv[optind+2]) || (Miss = atoc (argv[optind+2])) < 0)
			ERRMSG1 (misses (%s) is not a valid count, argv[optind+2])
		if (!isinteger (argv[optind+3]) || (Correct_rejection = atoc (argv[optind+3])) < 0)
			ERRMSG1 (correct-rejections (%s) is not a valid count, argv[optind+3])
		rawtable ();
		}
	computable ();
	exit (0);
	}

#define	dp_str(string) printf ("%10s", string)
#define	dp_real(value) printf ("%10.3f", value)
#define	dp_cnt(value)  printf ("%10ld", value)

rawtable ()
	{
	dp_str ("");     
		dp_str ("signal"); dp_str ("noise");
	putchar ('\n');
	dp_str ("yes");
		dp_cnt (Hit);  dp_cnt (False_alarm);       dp_cnt (Hit+False_alarm);
	putchar ('\n');
	dp_str ("no");
		dp_cnt (Miss); dp_cnt (Correct_rejection); dp_cnt (Miss+Correct_rejection);
	putchar ('\n');
	dp_str ("");
		dp_cnt (Hit+Miss); dp_cnt (False_alarm+Correct_rejection);
		dp_cnt (Hit+Miss+False_alarm+Correct_rejection);
	putchar ('\n');
	putchar ('\n');
	Hit_rate = Hit ? Hit / (double) (Hit + Miss) : 0.0;
	Precision = Hit ? (Hit / (double) (Hit + False_alarm)) : 0.0;
	False_alarm_rate = False_alarm
		? False_alarm / (double) (False_alarm + Correct_rejection)
		: 0.0;
		
	}

computable ()
	{
	double	zhr, zfar;        /* critical Z's for hit-rate & false-alarm-rate */
	double	dprime, beta;     /* computed statistics */
	double	criterion;        /* point above middle of noise distribution */
	if (Hit_rate <= 0.0 || Hit_rate >= 1.0)
		ERRMSG1 (The hit-rate (%g) must be greater than 0.0 and less than 1.0, Hit_rate)
	if (False_alarm_rate <= 0.0 || False_alarm_rate >= 1.0)
		ERRMSG1 (The false-alarm-rate (%g) must be greater than 0.0 and less than 1.0, False_alarm_rate)
	zhr    = critz (Hit_rate);
	zfar   = critz (False_alarm_rate);
	dprime = zhr - zfar;
	criterion = -zfar;
	beta   = normaldensity (zhr) / normaldensity (zfar);
	dp_str ("");
	dp_str ("hr");
	dp_str ("far");
	dp_str ("dprime");
	dp_str ("beta");
	dp_str ("criterion");
	if (Doprecision)
		dp_str ("precision");
	putchar ('\n');
	dp_str ("");
	dp_real (Hit_rate);
	dp_real (False_alarm_rate);
	dp_real (dprime);
	dp_real (beta);
	dp_real (criterion);
	if (Doprecision)
		dp_real (Precision);
	putchar ('\n');
	}

yesno (string) char *string;
	{
	int 	val;
	if (isinteger (string))
		{
		val = atoi (string);
		if (val == 1 || val == 0)
			return (val);
		}
	else /* check for upper/lower case strings */
		{
		if (!strcmp (string, "yes"))      return (1);
		if (!strcmp (string, "YES"))      return (1);
		if (!strcmp (string, "no"))       return (0);
		if (!strcmp (string, "NO"))       return (0);
		if (!strcmp (string, "signal"))   return (1);
		if (!strcmp (string, "SIGNAL"))   return (1);
		if (!strcmp (string, "noise"))    return (0);
		if (!strcmp (string, "NOISE"))    return (0);
		}
	ERRMSG1 (illegal value (%s) in input, string)
	/*NOTREACHED*/
	}

/*FUNCTION initial: returns local version of optind, index to first operand */
int
initial (argc, argv) char **argv;
	{
	extern	char *optarg;    /* option value accessed through this by getopt */
	extern	int  optind;     /* will be index to first operand */
	int 	opterr = 0;      /* count of number of errors */
	int 	flag;            /* option flag characters read in here */
	int 	nargs;           /* number of non-option arguments on cmd line */
	char	*optstring =     /* getopt string to be filled in */
		"LOVp";
	char	*usage =         /* part of usage summary to match optstring */
"[-p] [hit-rate false-alarm-rate] [hits false-alarms misses correct-rejections]";
	while ((flag = getopt (argc, argv, optstring)) != EOF)
		switch (flag)
			{
			default:
				opterr++;
				break;
			/* put option cases here */
			case 'p': Doprecision = TRUE; break;
			case 'O': InfoOptions = TRUE; break;
			case 'V': InfoVersion = TRUE; break;
			case 'L': InfoLimits = TRUE; break;
			}

	if (opterr) /* print usage message and bail out */
		{
		fprintf (stderr, "Usage: %s %s\n", argv[0], usage);
		exit (1);
		}
	
	usinfo ();

	nargs = argc - optind;
	if (nargs != 0 && nargs != 2 && nargs != 4)
		USAGE (usage)

	return (optind);
	}

usinfo ()
	{
	if (InfoVersion)
		pver (Version);
	if (InfoLimits)
		{
		plim (Argv0);
		const (MAXCHARS, "maximum number of characters in lines");
		}
	if (InfoOptions)
		{
		ppgm (Argv0, Purpose);
		lopt ('p',         "request computation of precision", Doprecision);
		}
	if (InfoVersion || InfoLimits || InfoOptions)
		exit (0);
	}
